User Update Audit

mkangia opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Tech Spec](https://docs.google.com/document/d/1Z5ZOu7O-BVOh62oCTc1JjNRsBAiMzs9i8Cf48q_9BDQ/edit#heading=h.pemmj11m3ymn)

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [ ] Risk label is set correctly
- [ ] All migrations are backwards compatible and won't block deploy
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [ ] If QA is part of the safety story, the "Awaiting QA" label is used
- [ ] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

mkangia committed:
update calls to log_model_change for

1. pass domain
2. always pass couch users

mkangia committed:
add model and update log_model_change accordingly

mkangia committed:
update current tests related to LogEntry

mkangia committed:
update current tests

mkangia committed:
add a happy path test

mkangia committed:
update some more tests

mkangia committed:
pass couch user for create calls

mkangia committed:
fix delete for user api

mkangia committed:
fix some more tests

mkangia committed:
fix more tests

mkangia committed:
fix calls to log_user_role_update to pass role

mkangia committed:
fix circular import

mkangia committed:
fix tests

mkangia committed:
change change_message to message in tests

mkangia committed:
fix tests

snopoke commented on corehq/apps/users/models.py:
```diff
@@ -1185,16 +1185,17 @@ def user_session_data(self):
         return session_data
 
-    def delete(self, deleted_by, deleted_via=None):
+    def delete(self, domain, deleted_by, deleted_via=None):
```
can't you get the domain from the object itself?

mkangia replied:
For a CommCareUser you could, but not for a web user because they could have multiple domain memberships?

snopoke commented on corehq/apps/users/models_sql.py:
```diff
@@ -179,6 +184,41 @@ class RoleAssignableBy(models.Model):
     )
 
 
+class UserUpdateMeta(jsonobject.JsonObject):
```
why is this a sub model an not just part of the main details object?

mkangia replied:
Sorry, i might have misunderstood how JsonObject is used then.

the SQL model has `details = JSONField(default=dict)`, which would be populated by wrapping using
```
class UpdateDetails(jsonobject.JsonObject):
    meta = jsonobject.ObjectProperty(UserUpdateMeta)
    changes = jsonobject.DictProperty()

class UserUpdateMeta(jsonobject.JsonObject):
    updated_via = jsonobject.StringProperty()
```

The `UserUpdateMeta` is kept separate to allow adding more meta properties in future. May be not a great example but like IP address of the request or user API version.

snopoke replied:
Why can't you just add more properties to the `UpdateDetails` directly? I just don't see the value of pulling it into a nested model.

mkangia replied:
updated to have "updated_via" on UpdateDetails itself.

snopoke commented on corehq/apps/users/models_sql.py:
```diff
@@ -179,6 +184,41 @@ class RoleAssignableBy(models.Model):
+        (UPDATE, _('Update')),
+        (DELETE, _('Delete')),
+    )
+    domain = models.CharField(max_length=255, db_index=True)
```
indexes should be added via the class `Meta` to avoid creating search indexes which aren't needed

snopoke commented on corehq/apps/users/models_sql.py:
```diff
@@ -179,6 +184,41 @@ class RoleAssignableBy(models.Model):
+    by_user_id = models.CharField(max_length=128, db_index=True)
+    details = JSONField(default=dict)  # UpdateDetails
+    message = models.TextField(_('change message'), blank=True)
+    action_time = models.DateTimeField(_('action time'), auto_now_add=True, editable=False)
+    action_flag = models.PositiveSmallIntegerField(_('action flag'), choices=ACTION_FLAG_CHOICES)
```
nit: rename these to `time` (or `date`) and `flag` respectively

mkangia replied:
used the standard "created_at" and then just "action" for the action.

snopoke commented on corehq/apps/users/models_sql.py:
```diff
@@ -179,6 +184,41 @@ class RoleAssignableBy(models.Model):
+    object_id = models.CharField(max_length=128, db_index=True)
+    by_user_id = models.CharField(max_length=128, db_index=True)
+    details = JSONField(default=dict)  # UpdateDetails
+    message = models.TextField(_('change message'), blank=True)
```
why are these verbose names necessary? Do they appear anywhere on the UI? They also don't seem to offer much more than the field name other than potential translation.

mkangia replied:
I am still fine tuning this. First step was to make the current implementation work with the new model, which i am still working on and hence the "message" field. I am not sure if it will be retained eventually or not, once details is populated with user fields and meta information is moved to it as well, this should not be needed.
For now, just picked the definition of "message" from django's LogEntry.

mkangia replied:
verbose names removed.

snopoke commented on corehq/util/model_log.py:
```diff
@@ -1,27 +1,25 @@
-    if isinstance(user, CouchUser):
-        user = user.get_django_user()
+    if not domain:
+        raise ValueError("Please pass domain")
```
why the special treatment for `domain`? None of the other fields are validated.

mkangia replied:
This is to catch cases where we don't pass in the domain internally. We should be allowing very specific cases like system updates, or invitation-less registrations to allow domain less log entries.

snopoke commented on corehq/apps/users/models_sql.py:
```diff
@@ -179,6 +184,41 @@ class RoleAssignableBy(models.Model):
+    changes = jsonobject.DictProperty()
+
+
+class HQLogEntry(models.Model):
```
This `models_sql` module is temporary for the duration of the couch-sql roles migration after which the models will be moved to the main `models` module. I think it would be better to put these directly in the `models` module.

snopoke commented on corehq/util/model_log.py:
```diff
@@ -35,11 +33,12 @@ def log_model_change(user, model_object, message=None, fields_changed=None, acti
+        object_type=model_object.doc_type,
+        object_id=model_object.get_id,
+        by_user_id=user.get_id,
+        details=UpdateDetails.wrap({}),
```
TODO: add details

snopoke commented on corehq/apps/users/models_sql.py:
```diff
@@ -179,6 +184,41 @@ class RoleAssignableBy(models.Model):
+    object_type = models.CharField(max_length=255, db_index=True, choices=(
+        ('CommCareUser', 'CommCareUser'),
+        ('WebUser', 'WebUser'),
+    ))
```
Is there a need to add the choices in here? They are only really useful for validation and adding new choices would require creating a migration.

mkangia committed:
allow domain less log entry

mkangia committed:
allow domain less entries for create where domain not available

mkangia commented on corehq/apps/cleanup/tasks.py:
```diff
@@ -253,7 +253,8 @@ def delete_web_user():
             if web_user:
-                web_user.delete(deleted_by=None, deleted_via=__name__)
+                # ToDo: Should this pass "system" as the user? Currently won't be logged at all
+                web_user.delete(None, deleted_by=None, deleted_via=__name__)
```
Cool, updating to  `__name__ + '.delete_web_user'`.

mkangia commented:
thanks for adding @calellowitz to this @orangejenny ,

Just flagging that this is not ready for a good review, but good enough to drop early thoughts. I did offline with Simon about this before pinging him for just quick thoughts. So it's okay to not worry about reviewing this yet unless you are curious to know more :)

mkangia committed:
track method path instead

mkangia committed:
Merge branch 'master' into mk/user-history-log

mkangia committed:
fix tests

mkangia committed:
allow system user to update users without domain

mkangia committed:
isort + unused imports

mkangia committed:
keep it simple

mkangia committed:
add ToDo

mkangia committed:
update model

- remove verbose names
- add indexes via Meta, skip object type index
- rename action_flag to action
- rename action_time to created_at
- remove object type choices

mkangia committed:
resolve conflicts; merge master

mkangia committed:
missed during conflicts, pass couch user

mkangia committed:
pass couch user now

mkangia committed:
update to new name

mkangia committed:
update log method to

- make message optional
- accept changed_via to avoid using message

mkangia committed:
remove unused import

mkangia committed:
fix documentation

mkangia committed:
save user doc on create/delete

mkangia committed:
track role updates

mkangia committed:
treat unretire like a create

mkangia committed:
reorder for better read

SYSTEM user is always allowed domain less log

snopoke commented on corehq/util/model_log.py:
```diff
@@ -1,45 +1,56 @@
-from django.utils.encoding import force_text
+from corehq.apps.users.models_sql import HQLogEntry, UpdateDetails
+
+MODEL_PROPERTIES_TO_IGNORE = ['_id', '_rev', 'reporting_metadata', 'password']
```
add `devices` and `last_device` and `device_ids`

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29805/commits/faa0d255a0a42b791a8efd11e3c14efe2c926ec9

mkangia commented on corehq/apps/users/models.py:
```diff
@@ -1436,7 +1437,7 @@ def from_django_user(cls, django_user, strict=False):
     @classmethod
     def create(cls, domain, username, password, created_by, created_via, email=None, uuid='', date='',
-               first_name='', last_name='', metadata=None, **kwargs):
+               first_name='', last_name='', metadata=None, can_skip_domain=False, **kwargs):
```
I added this here to reflect the change in definition of CommCareUser/WebUser that inherit from CouchUser. It's used and passed there. We could also let it get consumed in kwargs if this is confusing.

snopoke replied:
I don't see it being passed from either of those methods to the 'super' method (this method). If it's unused here then you shouldn't add it in as a kwarg unless it's part of a duck typed interface (which it isn't in this case).

mkangia replied:
Ah, i missed passing it in super.

> unless it's part of a duck typed interface (which it isn't in this case)

Makes sense. I will remove it from here.

mkangia commented on corehq/util/model_log.py:
```diff
@@ -1,45 +1,56 @@
 
-    if message is not None and fields_changed is not None:
-        raise ValueError("'message' and 'fields_changed' are mutually exclusive")
+    if not domain and user != SYSTEM_USER_ID and not can_skip_domain:
```
We could though that would make it generic.

Its just that, I was trying to ensure we log domain in all places where its possible to do so, so that all relevant log entries are visible in the report when we filter by domain. If there is a generic check like that, its possible domain is not passed at times when available, leading to domain less logs, though there are very few places where a web user is registered, but I didn't want to assume that i knew all. 

And plus there is also https://github.com/dimagi/commcare-hq/blob/99c8a0231dca39f5a32e78c85730e48c69611eab/corehq/apps/sso/backends.py#L112 where domain might not be available

mkangia committed:
add more properties to ignore

mkangia commented on corehq/apps/users/models.py:
```diff
@@ -1725,12 +1725,12 @@ def save(self, fire_signals=True, spawn_task=False, **params):
             sync_usercases_if_applicable(self, spawn_task)
 
-    def delete(self, deleted_by, deleted_via=None):
+    def delete(self, domain, deleted_by, deleted_via=None):
```
Just to confirm,
This is because a "domain" would be present on the `CommCareUser` itself which it is bound to.

Just want to confirm if things are now different because of Mirrored domains? not sure if that is the right term to put it. Is it possible for a domain to delete a commcare user bound to a different domain?

orangejenny replied:
It either is or likely will soon be possible for one domain to delete a different domain's mobile worker, yes.

mkangia replied:
cool. Thanks @orangejenny 

@snopoke let me know if it sounds reasonable now to pass along domain for both mobile and web users.

mkangia committed:
remove redundant argument

snopoke commented on corehq/util/model_log.py:
```diff
@@ -1,45 +1,59 @@
+        object_type=model_object.doc_type,
+        object_id=model_object.get_id,
+        by_user_id=SYSTEM_USER_ID if user == SYSTEM_USER_ID else user.get_id,
+        details=UpdateDetails.wrap({
+            'changes': _get_change_details(model_object, action, fields_changed),
+            'changed_via': changed_via,
+        }),
```
I don't see much value in wrapping this here since you're just making a dict
```suggestion
        details={
            'changes': _get_change_details(model_object, action, fields_changed),
            'changed_via': changed_via,
        },
```

snopoke commented on corehq/apps/users/models_sql.py:
```diff
@@ -228,6 +233,41 @@ class RoleAssignableBy(models.Model):
+            models.Index(fields=['domain']),
+            models.Index(fields=['object_id'])
+        ]
+        index_together = ["domain", "by_user_id"]
```
why is this a combined index and the other one isn't? You're effectively creating two `domain` indexes.
